crates/generate/src/funcs.rs

```diff
@@ -361,6 +363,7 @@ where
             witx::BuiltinType::String => unimplemented!("string types"),
         },
         witx::Type::Enum(_) | witx::Type::Flags(_) | witx::Type::Int(_) => write_val_to_ptr,
-        _ => unimplemented!("marshal result"),
+        witx::Type::Struct(s) if struct_is_copy(&s) => write_val_to_ptr,
```

Comment:
Is there anything stopping a clone struct from being returned? If we can construct a test case to exercise it, I'd prefer to allow non-copy structs to be returned as well, I don't see any issue for it in this particular code.

Reply:
So I wasn't sure about that one. If we can return a non-copy struct, that would imply the possibility of returning a struct with pointers in it. Does that mean we alloc some memory inside an interface func, populate it (or not, that's pretty optional here), and then return it? I guess what I'm asking here, are we allowed/expected to do that?

Reply:
OK, I think I get it now. We don't necessarily need to be able to allocate in guest's memory. We could in principle wrap `GuestPtr`s in a struct which we'd then return. Am I on the right track with this one?

Reply:
OK, so allowing for non-copy structs will require slightly different impl than simply reusing `write_val_to_ptr` since we won’t necessarily have `as_ref_mut()` for types which are only `GuestTypeClone`. In particular, I’m thinking here of `GuestPtr` for instance.

Another problem I see is in the test case. I figured, the simples test case would be simply to construct `PairIntPtrs` struct from two input `GuestPtr`s but that causes lifetime problems which I’m guessing is because we’re mainly relying on anonymous lifetimes when spawning the trait from `witx`.

Reply:
@pchickey Huzzah, managed to return a non-copy struct! Have a look! FYI, I've had to rebase to #20 since two changes to the generator were needed to make this happen:
1. since now it is possible to return a struct which can have a lifetime attached to it, we need to annotate the interface function with the said *explicit* lifetime as well; in such cases, anonymous lifetimes won't cut it unfortunately
2. since a struct can be non-copy, I needed #20 which ensures `Copy => Clone` implication, and I needed to change return value marshaling to not rely on `as_ref_mut` which *is not* available if only `Clone`, and instead resort to `write_ptr_to_guest` which now is always available (due to the said implication on copyability).